assembl/tasks/watson.py

```diff
@@ -0,0 +1,175 @@
+                computation.status = "success"
+                for keyword in result.get('keywords', ()):
+                    tag = Tag.getOrCreateTag(
+                        keyword['text'], lse.locale, post.db)
+                    tag.simplistic_unify(translator)
+                    post.db.add(PostKeywordAnalysis(
+                        post=post, source=computation,
+                        tag=tag, score=keyword['relevance']))
+                for category in result.get('categories', ()):
+                    tag = Tag.getOrCreateTag(
+                        category['label'], lse.locale, post.db)
+                    tag.simplistic_unify(translator)
+                    post.db.add(PostKeywordAnalysis(
+                        post=post, source=computation, category=True,
+                        tag=tag, score=category['score']))
+                for concept in result.get('concepts', ()):
+                    dbconcept = DBPediaConcept.get_or_create(
+                        concept['dbpedia_resource'], post.db)
+                    dbconcept.identify_languages(desired_locales, post.db)
+                    post.db.add(PostDBPediaConceptAnalysis(
+                        post=post, source=computation,
+                        value=dbconcept, score=keyword['relevance']))
+                sentiments = {}
+                if result.get('emotion', None):
+                    emotion = result['emotion']['document']['emotion']
+                    sentiments.update(dict(
+                        anger=emotion['anger'],
+                        disgust=emotion['disgust'],
+                        fear=emotion['fear'],
+                        joy=emotion['joy'],
+                        sadness=emotion['sadness'],
+                    ))
+                if result.get('sentiment', None):
+                    sentiments['sentiment'] = \
+                        result['sentiment']['document']['score']
+                if sentiments:
+                    post.db.add(PostWatsonV1SentimentAnalysis(
+                        post=post,
+                        source=computation,
+                        text_length=len(result['analyzed_text']),
+                        **sentiments
+                    ))
+            except Exception as e:
+                computation.result = str(e)
+                computation.status = "failure"
+                computation.retries = (computation.retries or 0) + 1
+
+
+def get_or_create_computation_on_post(post, process_name, features):
+    from ..models.computation import ComputationOnPost, ComputationProcess
+    for computation in post.computations:
+        if (computation.process.name == process_name and
+                computation.features == features):
+            return computation
+    process = ComputationProcess.by_name(process_name, post.db)
+    computation = ComputationOnPost(
+        post=post, process=process,
+        parameters=features)
+    post.db.add(computation)
```

Comment:
Then this function `get_or_create_computation_on_post`, is incomplete. It's designed to work under Celery, where it's wrapped in a transaction. But the function is not as useful in it's own context. Perhaps pass the transaction `in_tx=False` to the function, so that it can be used outside of a transaction as well. Agree?

Reply:
use process_post_watson